app/models/concerns/course/discussion/post/ordering_concern.rb

```diff
@@ -0,0 +1,44 @@
+module Course::Discussion::Post::OrderingConcern
+  extend ActiveSupport::Concern
+
+  # Sorts all posts in a collection in topographical order.
+  #
+  # By convention, each post is represented by an array. The first element is the post itself,
+  # the second is the children of the array.
+  class PostSort
+    include Enumerable
+    delegate :each, to: :@sorted
+
+    # Constructor.
+    #
+    # @param [Array<Course::Discussion::Post>] posts The posts to sort.
+    def initialize(posts)
+      @posts = posts
+      @sorted = sort(nil)
+    end
+
+    private
+
+    def sort(post)
+      children_posts = @posts.select { |child_post| child_post.parent == post }
+      children_posts.map do |child_post|
+        [child_post].push(sort(child_post))
+      end
+    end
+  end
+
+  module ScopeMethods
+    # Returns a set of recursive arrays indicating the parent-child relationships of posts.
+    #
+    # @return [Enumerable]
+    def ordered_topologically
+      PostSort.new(self)
+    end
+  end
+
+  included do
+    scope :ordered_topologically, (lambda do
+      ScopeMethods.instance_method(:ordered_topologically).bind(current_scope).call
```

Comment:
@lowjoel I don't understand why not just adopt the TSort on current_scope ?  instead you are doing the binding and delegation thing, it's complex...